### PR TITLE
ESQL: Use faster field caps

### PR DESCRIPTION
**Pull request by nik9000**
The field capabilities has an internal-only option to produce un-merged output. This expands that option to be available to any caller inside of Elasticsearch and uses it in ES|QL to speed up queries across many indices with many fields. Across 5,000 indices with a couple thousand fields each (metricbeat) the `FROM *` query went from 600ms to 60ms. Across 50,000 indices that went from 6600ms to 600ms. 600ms is still too slow for such a simple query, but one step at a time!

This is faster because field capabilities wants to present a field-centric result but ES|QL actually needs a different flavor of field-centric result with some differences smoothed away. If we take over the merging process we can use a few tools that the field caps API uses internally to be fast - mostly the sha256 of the mapping - to save on doing work that wasn't available in the other view. Also, two merges is more expensive than one.

That 90% reduction in runtime doesn't banish field caps from the flamegraphs. You still see it, but it's now much less prominent. And you don't see the merging process at all. Now it's all data-node side operations field caps.

Relates to #103369

### CODE REVIEW

File: x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/plugin/EsqlPlugin.java [hunk 1]
```diff
@@ -102,6 +103,10 @@ public Collection<?> createComponents(PluginServices services) {
                     services.clusterService().getClusterName().value(),
                     EsqlDataTypeRegistry.INSTANCE,
                     Set::of
+                ),
+                new EsqlIndexResolver(
+                    services.client(),
+                    EsqlDataTypeRegistry.INSTANCE
```
Comment (nik9000) [inline x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/plugin/EsqlPlugin.java:109]: I'm only keep the original IndexResolver so I can compare the results.

File: x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/session/EsqlSession.java [hunk 2]
```diff
@@ -197,18 +203,97 @@ private <T> void preAnalyzeIndices(LogicalPlan parsed, ActionListener<IndexResol
...
-                listener,
-                EsqlSession::specificValidity,
-                IndexResolver.PRESERVE_PROPERTIES,
-                // TODO no matter what metadata fields are asked in a query, the "allowedMetadataFields" is always _index, does it make
-                // sense to reflect the actual list of metadata fields instead?
-                IndexResolver.INDEX_METADATA_FIELD
-            );
+            esqlIndexResolver.resolveAsMergedMapping(table.index(), fieldNames, false, Map.of(), listener);
+
+//            indexResolver.resolveAsMergedMapping(table.index(), fieldNames, false, Map.of(), new ActionListener<>() {
```
Comment (nik9000) [inline x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/session/EsqlSession.java:208]: I've been swapping the line above with these lines to make sure I output the same result.

**Review by nik9000 — COMMENTED** [review]
I imagine it'd be nice to port all of the resolution infrastructure to un-merged results but there is lot of it, and a lot of layers that I don't understand. So I did the simplest thing I could think of - I made my own version that i could be sure was fairly efficient and then asserted it had the same output as the original.

**Review by astefan — COMMENTED** [review]
I really like those numbers in performance improvements and it seems we are going in the right direction.

But, the most scary part about the IndexResolver removal is that it has some years of maintenance and fixes and some of the tests that are indirectly using it live in other projects. For example, one of the most important test suites I've created that deals with many edge cases is `FieldExtractorTestCase` in ES SQL.

If we are to use our own flavor of the IndexResolver (and I am in favor of this given the preliminary performance gains) we'd need to port some of the existent tests in SQL and QL projects here, in ESQL.
`isAlias` is being used to build the hierarchy of fields when the fieldcaps response is not telling us if the type of a field is actually an `alias` type. **It is possible this was useful to ES SQL to extract data from _source when the `fields` API didn't exist** but I am not 100% sure it is still needed now (`FieldExtractorTestCase.testAliasFromDocValueField` is a test that checks this type of hierarchy).

**Comment by nik9000** [comment]
I talked to @astefan and the plan now is to try and port some tests from SQL that have more complex mappings to ESQL. And to run both resolution mechanisms if assertions are enabled - basically to uncomment that big testing blob I have. He'll do a little more research on his side so he can review this in more depth. I'll port the test.

**Comment by nik9000** [comment]
> For example, one of the most important test suites I've created that deals with many edge cases is `FieldExtractorTestCase` in ES SQL.

I've ported it!

**Comment by nik9000** [comment]
@astefan, I've ported the tests from SQL and everything *seems* to be passing. Could you have another look?

**Review by astefan — COMMENTED** [review]
I think the tests are fine in `FieldExtractorTestCase` 👍 

I started digging and trying out different things. I looked at `IndexResolverTests` from sql and I think we should port that one as well (in that one we have a text for fix in https://github.com/elastic/elasticsearch/issues/100869 for example). For one of those tests I tried out - `testMergeIncompatibleCapabilities` - I get a `ql and esql didn't make the same resolution: types differ [INTEGER] != [UNSUPPORTED]` message.
It is because `doc_values: false`, which means not aggregateable in one of the indices.

Going further and ignoring the awesome EsqlIndexResolver-IndexResolver checks you added in EsqlSession, for that specific mapping in `testMergeIncompatibleCapabilities` and query `from test* | stats max(emp_no)` I get `Cannot use field [emp_no] due to ambiguities being mapped as aggregatable except in [test2]` which is ok..., BUT we have no tests for these types of errors :-(.

So:
- `IndexResolverTests` needs some of its tests ported to ESQL. I am not particularly fond of this class because for some tests it's loading the mappings "manually" (not actually relying on field_caps natural output) and then asserting behavior on them
- we need some coverage for these mapping incompatibilities that reach the planning (the Analyzer in this case). I'll look into this today

File: x-pack/plugin/esql/qa/server/src/main/java/org/elasticsearch/xpack/esql/qa/rest/FieldExtractorTestCase.java [hunk 3]
```diff
@@ -0,0 +1,941 @@
...
+        assertMap(
+            result,
+            matchesMap().entry("columns", List.of(columnInfo("text_field", "text"), columnInfo("text_field.ip", "ip")))
+                .entry("values", List.of(matchesList().item(value).item(null)))
+        );
+    }
+
+    /**
+     * <pre>
+     * "int_field": {
```
[the remark this answers is not in the document]
Comment (nik9000) [reply x-pack/plugin/esql/qa/server/src/main/java/org/elasticsearch/xpack/esql/qa/rest/FieldExtractorTestCase.java:395]: I'll clean this one up. I always write `int` instead of `integer` when I work with that mapper. Not sure why. OTOH, I tend to use `str` when I mean `text|keyword`. Old habit. I'll double check it.

**Comment by nik9000** [comment]
> It is because `doc_values: false`, which means not aggregateable in one of the indices.

That's interesting! I think ESQL is fine aggregating fields that return `false` from `aggregatable` - that's mostly a thing for `_search`. ESQL will just load the field from `_source`.

**Comment by nik9000** [comment]
> That's interesting! I think ESQL is fine aggregating fields that return `false` from `aggregatable` - that's mostly a thing for `_search`. ESQL will just load the field from `_source`.

But let's not change behavior in this PR. I'll make sure it all lines up. We can change behavior later.

**Comment by nik9000** [comment]
> Nevertheless, some of the tests (adapted to ESQL) in IndexResolverTests should be ported.

I'm on it!

**Comment by nik9000** [comment]
@astefan I've ported the tests in `IndexResolverTests` to the big integration test. I'm cleaning up a few remaining test failures, but I think this is ready for another round.

File: x-pack/plugin/esql/qa/server/src/main/java/org/elasticsearch/xpack/esql/qa/rest/FieldExtractorTestCase.java [hunk 4]
```diff
@@ -596,6 +630,270 @@ public void testByteFieldWithIntSubfieldTooBig() throws IOException {
...
+        index("test", """
+            {"f": "192.168.0.1/24"}""");
+
+        ResponseException e = expectThrows(
+            ResponseException.class,
+            () -> runEsqlSync(new RestEsqlTestCase.RequestObjectBuilder().query("FROM test* | SORT f, f.raw | LIMIT 3"))
+        );
+        String err = EntityUtils.toString(e.getResponse().getEntity());
+        assertThat(err, containsString("Cannot use field [f] with unsupported type [ip_range]"));
+        assertThat(err, containsString("Cannot use field [f.raw] with unsupported type [ip_range]"));
```
Comment (astefan) [inline x-pack/plugin/esql/qa/server/src/main/java/org/elasticsearch/xpack/esql/qa/rest/FieldExtractorTestCase.java:794]: I don't remember why we marked the hierarchy of unsupported fields as unsupported as well. My best quess is that, since we "materialized" the values of subfields from the parent fields (because this is what actually gets indexed as part of the `_source`), and parents had types that couldn't be displayed by SQL, we set the entire hierarchy as unsupported. For the record, [this PR](https://github.com/elastic/elasticsearch/pull/50823/files) added this behavior. @costin do you remember why we did this for SQL? Is my guess above correct?

Nevertheless, is this "limitation" still valid for ESQL who reads the values from Lucene differently than what it used to be for SQL?

For example, this use case does seem to make sense for ESQL, unless I am forgetting/missing something:

test1

```
            "file": {
                "type": "ip_range",
                "fields": {
                    "raw": {
                        "type": "keyword"
                    }
                }
            }
```

test2

```
            "file": {
                "type": "text",
                "fields": {
                    "raw": {
                        "type": "keyword"
                    }
                }
            }
```

For query `from test*` we return `file.raw` as `unsupported`.

Nik's test above replicates what happens in `main`, but `main` seems to be broken from at least one angle: `file.raw` is not of type `ip_range` as the error message says.

--- 

### In conclusion:
- should hierarchies of unsupported field types be also marked as unsupported (current behavior in SQL and old IndexResolver)? If so, why?
- there is some context related to this specific use case:
  - for SQL we did this change https://github.com/elastic/elasticsearch/pull/50823/files but **it may be only SQL specific**
  - we already changed the SQL's default behavior slightly with this PR https://github.com/elastic/elasticsearch/pull/100875/files where we recognized that ESQL is different from SQL in one matter

**Review by astefan — COMMENTED** [review]
@nik9000 thank you for digging deeper in the IndexResolver area and for adding more tests. I have added a comment that is not strictly related to this PR but to understand and hopefully simplify ESQL's handling of unsupported field types.

File: x-pack/plugin/esql/qa/server/single-node/src/javaRestTest/java/org/elasticsearch/xpack/esql/qa/single_node/RestEsqlIT.java [hunk 5]
```diff
@@ -185,8 +185,7 @@ public void testIncompatibleMappingsErrors() throws IOException {
         assertException("from test_alias | where _size is not null | limit 1", "Unknown column [_size]");
         assertException(
             "from test_alias | where message.hash is not null | limit 1",
-            "Cannot use field [message.hash] due to ambiguities",
-            "incompatible types: [integer] in [index2], [murmur3] in [index1]"
+            "Cannot use field [message.hash] with unsupported type [murmur3]"
```
Comment (nik9000) [inline x-pack/plugin/esql/qa/server/single-node/src/javaRestTest/java/org/elasticsearch/xpack/esql/qa/single_node/RestEsqlIT.java:188]: I'm not sure how this works, but I don't think it's a killer for the approach, but we should decide. This example makes it look like using two field types will give you the ambiguities error. But in other examples the QL index resolution stuff will give you the unsupported error. We could do either one on the ESQL side, but I picked doing the `unsupported` error because it's the first such error message I found. I'd be happy to switch it, but that'll break other tests which I'll have to coax back to grene.

Comment (astefan) [reply x-pack/plugin/esql/qa/server/single-node/src/javaRestTest/java/org/elasticsearch/xpack/esql/qa/single_node/RestEsqlIT.java:188]: These tests here are new (added with https://github.com/elastic/elasticsearch/pull/105544/files) because we didn't cover this in tests.

In any case, my personal opinion is that your approach is the better approach: tell the user early that there is field that's unsupported, rather than telling "these two fields have mixed incompatible types" and the user then trying to make them compatible (reindexing or whatever) and later to find out that the effort was in vain because the type they chose to make compatible is the wrong type.

Comment (nik9000) [reply x-pack/plugin/esql/qa/server/single-node/src/javaRestTest/java/org/elasticsearch/xpack/esql/qa/single_node/RestEsqlIT.java:188]: I'm glad you like it!

**Comment by nik9000** [comment]
I've talked with @andreabucci and I'm going to port `VerifierErrorMessagesTests` - at least, as much of it as applies.

**Comment by nik9000** [comment]
> I've talked with @andreabucci

sorry, wrong @. I just typed `@and` and hit enter. Sorry!

File: x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/execution/PlanExecutor.java [hunk 6]
```diff
@@ -29,14 +30,16 @@
...
     private final IndexResolver indexResolver;
+    private final EsqlIndexResolver esqlIndexResolver;
     private final PreAnalyzer preAnalyzer;
     private final FunctionRegistry functionRegistry;
     private final Mapper mapper;
     private final Metrics metrics;
     private final Verifier verifier;
 
-    public PlanExecutor(IndexResolver indexResolver) {
+    public PlanExecutor(IndexResolver indexResolver, EsqlIndexResolver esqlIndexResolver) {
```
Comment (costin) [inline x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/execution/PlanExecutor.java:40]: Why keep IndexResolver around? To check the behavior is the same?

File: x-pack/plugin/esql/qa/server/src/main/java/org/elasticsearch/xpack/esql/qa/rest/FieldExtractorTestCase.java (hunk 4 above)
Comment (costin) [reply x-pack/plugin/esql/qa/server/src/main/java/org/elasticsearch/xpack/esql/qa/rest/FieldExtractorTestCase.java:805]: Great points @astefan . Tl;dr is that ESQL already has and will have better handling of field extraction and thus conflict resolution than SQL and by having a different resolved, we should be able to evolve this better.
in SQL extraction is done through field caps hence why we expected a whole field hierarchy to match - it might not have been perfect but it got us pretty far.

In ESQL that assumption no longer holds - #100875 allowed subfields to allow the following [mapping](https://github.com/elastic/elasticsearch/pull/100875/files#diff-70d0893691f1bd1e3b4be7b7146fbbbb1168851e8cdcf9ace6c26e5b6839b1a9):

index 1 - `field.name` keyword
index 2 - `field` = keyword

In Index 1, field is object, as parent for field.name. This conflicts with field in index2 which is a keyword - using the SQL behavior we discarded field.name as unsupported. However there's no `field.name` in index2 and thus there's no mapping conflict - in other words, field is problematic but field.name is not.

I think we're in a good enough place with this and considering the performance savings I suggest we move forward with it - there's a chance there might be some subtle cases we haven't covered yet and by changing the behavior we're going to run into them but right now we have enough testing in place to be confident about them.
Furthermore I would backport this behavior in 8.13 to test it as much as possible. Potentially introduce an internal flag to have the fallback behavior for a while when running into problems.

**Review by costin — APPROVED** [review]
LGTM

File: x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/execution/PlanExecutor.java (hunk 6 above)
Comment (nik9000) [reply x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/execution/PlanExecutor.java:40]: Yup

**Comment by nik9000** [comment]
> Potentially introduce an internal flag to have the fallback behavior for a while when running into problems.

A flag conflicts with any behavior changes we do after this. At some point we'll want to diverge and then we'd have to be really careful with testing with the flag on and off. That's trick!

**Review by astefan — APPROVED** [review]
LGTM. I'm really happy with this PR. @nik9000 thank you for the patience and the performance improvements this change brings.

**Comment by nik9000** [comment]
> LGTM. I'm really happy with this PR. @nik9000 thank you for the patience and the performance improvements this change brings.

Thanks for your patience with me too!

**Comment by nik9000** [comment]
> The backport operation could not be completed due to the following error:

oh. Well. That makes sense. I'll do it myself.

**Comment by nik9000** [comment]
Too late to backport I think.
